runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -400,8 +398,35 @@ private void setTimer(WindowedValue<InputT> timerElement, TimerInternals.TimerDa
     }
   }
 
+  @SuppressWarnings("ByteBufferBackingArray")
+  private void fireCleanupTimers() {
+    while (!cleanupTimers.isEmpty()) {
+      InternalTimer<?, TimerInternals.TimerData> timer = cleanupTimers.remove();
+      final ByteBuffer encodedKey = (ByteBuffer) timer.getKey();
+      try {
+        // still need to process as timer, see CleanupTimer
+        stateBackendLock.lock();
```

Comment:
Consider only locking once before processing all the timers, but probably negligible.

Reply:
Probably, it would also require to first check that there is at least one timer before calling this method to not take the lock unnecessarily.